[HttpKernel] Show full URI when route not found

ruudk opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      |no
| New feature?  | yes
| Deprecations? | no
| License       | MIT

When accessing a route that does not exist, Symfony throws a `NotFoundHttpException` that says `No route found for "POST /path"`.

On some projects this might be good enough to find the root cause, but on projects that have lots of routes on different hosts, it becomes hard to understand how the request was initiated. Was it done over HTTP or HTTPS? What was the hostname? Did the user specify a port?

To make this easier, we now show the full URI of the path, like this: `No route found for "POST https://www.symfony.com/path"`.

nicolas-grekas commented on src/Symfony/Component/HttpKernel/CHANGELOG.md:
```diff
@@ -5,6 +5,7 @@ CHANGELOG
 ---
 
  * marked the class `Symfony\Component\HttpKernel\EventListener\DebugHandlersListener` as internal
+ * added full request uri when route is not found
```
no need for this line imho

ruudk replied:
Removed.

nicolas-grekas reviewed (commented):
can you please add a test case?

ruudk commented:
@nicolas-grekas Added some tests.

OskarStark commented on src/Symfony/Component/HttpKernel/Tests/EventListener/RouterListenerTest.php:
```diff
@@ -214,4 +218,61 @@ public function testRequestWithBadHost()
+
+        $context = new RequestContext();
+
+        $urlMatcher = $this->getMockBuilder(\Symfony\Component\Routing\Matcher\UrlMatcherInterface::class)
+            ->disableOriginalConstructor()
+            ->getMock();
```
```suggestion
        $urlMatcher = $this->createMock(\Symfony\Component\Routing\Matcher\UrlMatcherInterface::class);
```

plus, shouldn't we import the the class instead of writing the FQCN?

ruudk replied:
I agree, but the whole test is like this. And I copied it. Should I fix the whole file?

nicolas-grekas replied:
yes please

ruudk replied:
Fixed.

OskarStark commented on src/Symfony/Component/HttpKernel/Tests/EventListener/RouterListenerTest.php:
```diff
@@ -91,7 +99,7 @@ public function testRequestMatcher()
         $event = new RequestEvent($kernel, $request, HttpKernelInterface::MASTER_REQUEST);
 
-        $requestMatcher = $this->getMockBuilder(\Symfony\Component\Routing\Matcher\RequestMatcherInterface::class)->getMock();
+        $requestMatcher = $this->getMockBuilder(RequestMatcherInterface::class)->getMock();
```
Please use createMock instead

Same for the others

PS maybe a rebase is enough, I fixed a lot of them, maybe it's already done ✌🏻

ruudk replied:
@OskarStark Rebased and applied fixes. WDYT?

ruudk committed:
Show full URI when route not found

When accessing a route that does not exist, Symfony throws a `NotFoundHttpException` that says `No route found for "POST /path"`.

On some projects this might be good enough to find the root cause, but on projects that have lots of routes on different hosts, it becomes hard to understand how the request was initiated. Was it done over HTTP or HTTPS? What was the hostname? Did the user specify a port?

To make this easier, we now show the full URI of the path, like this: `No route found for "POST https://www.symfony.com/path"`.

OskarStark commented on src/Symfony/Component/HttpKernel/Tests/EventListener/RouterListenerTest.php:
```diff
@@ -217,4 +221,57 @@ public function testRequestWithBadHost()
+    public function testMethodNotAllowedException()
+    {
+        $this->expectException(MethodNotAllowedHttpException::class);
+        $this->expectExceptionMessage('No route found for "GET https://www.symfony.com/path": Method Not Allowed (Allow: POST)');
```
```suggestion
        $this->expectExceptionMessage('No route found for "GET https://www.symfony.com/path": Method Not Allowed (allowed methods: POST)');
```
?

ruudk replied:
not part of my change.

ro0NL replied:
it's HTTP talk for `405 Method Not Allowed`: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Allow

i'd keep it

fabpot commented:
Thank you @ruudk.
